tico/experimental/quantization/ptq/wrappers/fairseq/quant_mha.py

```diff
@@ -0,0 +1,391 @@
+
+        # scale & static causal mask
+        self.register_buffer(
+            "scale_const", torch.tensor(self.head_dim**-0.5), persistent=False
+        )
+        mask = torch.full((1, 1, max_seq, max_seq), float(self.mask_fill_value))
+        mask.triu_(1)
+        self.register_buffer("causal_mask_template", mask, persistent=False)
+
+        # observers (no *_proj_out here; PTQWrapper handles module outputs)
+        mk = self._make_obs
+        self.obs_query_in = mk("query_in")
+        self.obs_key_in = mk("key_in")
+        self.obs_value_in = mk("value_in")
+        self.obs_kpm_in = mk("kpm_in")
+        self.obs_causal_mask = mk("causal_mask")
+        self.obs_q_fold = mk("q_fold")
+        self.obs_k_fold = mk("k_fold")
+        self.obs_v_fold = mk("v_fold")
+        self.obs_scale = mk("scale")
+        self.obs_logits_raw = mk("logits_raw")
+        self.obs_logits = mk("logits_scaled")
+        self.obs_attn_mask_add = mk("obs_attn_mask_add")
+        self.obs_kp_mask_add = mk("obs_kp_mask_add")
+        self.obs_softmax = mk("softmax")
+        self.obs_attn_out = mk("attn_out")
+
+        safe_name = (
+            fp_name if (fp_name not in (None, "", "None")) else f"QuantFsMHA_{id(self)}"
+        )
+        assert safe_name is not None
+        self._state_key = safe_name + ".attn_state"
+
+    def _get_input_buffer(
+        self,
+        incremental_state: Optional[Dict[str, Dict[str, Optional[torch.Tensor]]]],
+    ) -> Optional[Dict[str, Optional[torch.Tensor]]]:
+        """Return saved KV/mask dict or None."""
+        if incremental_state is None:
+            return None
+        return incremental_state.get(self._state_key, None)
+
+    def _set_input_buffer(
+        self,
+        incremental_state: Optional[Dict[str, Dict[str, Optional[torch.Tensor]]]],
+        buffer: Dict[str, Optional[torch.Tensor]],
+    ):
+        """Store KV/mask dict in incremental_state."""
+        if incremental_state is not None:
+            incremental_state[self._state_key] = buffer
+        return incremental_state
+
+    # ---- utils ----
+    def _fold_heads(self, x: torch.Tensor, B: int) -> torch.Tensor:
+        # [T,B,E] -> [B*H, T, Dh]
+        T = x.size(0)
+        x = x.view(T, B, self.num_heads, self.head_dim).permute(1, 2, 0, 3).contiguous()
+        return x.view(B * self.num_heads, T, self.head_dim)
+
+    def _unfold_heads(self, x: torch.Tensor, B: int, T: int) -> torch.Tensor:
+        # [B*H, T, Dh] -> [T,B,E]
+        x = x.view(B, self.num_heads, T, self.head_dim).permute(2, 0, 1, 3).contiguous()
+        return x.view(T, B, self.embed_dim)
+
+    def _get_state(
+        self, incremental_state: Optional[Dict], create: bool = False
+    ) -> Optional[Dict]:
+        if incremental_state is None:
+            return None
+        if create and self._state_key not in incremental_state:
+            incremental_state[self._state_key] = {}
+        return incremental_state.get(self._state_key, None)
```

Comment:
Is this function in use? Looks like duplicate of _get_input_buffer

Reply:
Maybe I used this when debugging. Let's remove it.